Remove server from routingTable if failed to establish connection

zhenlineo opened the pull request:
* Fixed the bug where if the client failed to establish connection with the server, the server is not removed from the routing table
* Refactoring on routing table

Commit:
Fixed the bug where if the client failed to establish connection with the server the server is not removed from the routing table

Commit:
refactoring on routing table

Commit:
Fix the failing test due to unreliable error type on different platforms

Commit:
Fix the error code for ssl connections
If failing to auth server to client/client to server, then fail with clientException
If failing to establish connection with the server, then fail with ServiceUnavailableException
If the connects between the server and the client breaks, then fail with ServiceUnavailableException

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/cluster/ClusterRoutingTable.java:
```diff
@@ -0,0 +1,126 @@
+
+    private final Clock clock;
+    private long expirationTimeout;
+    private final RoundRobinAddressSet readers, writers, routers;
```
Could you please split this line into separate declarations?

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/cluster/ClusterRoutingTable.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Override
+    public synchronized HashSet<BoltServerAddress> update( ClusterComposition cluster )
```
It is better to have `Set<BoltServerAddress>` as the return type

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/cluster/ClusterRoutingTable.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Override
+    public boolean stale()
```
Maybe `isStale()`?

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/cluster/Rediscovery.java:
```diff
@@ -0,0 +1,95 @@
+
+    // Given a old routing table and connection pool, use the connection composition provider to obtain a new
+    // cluster composition, which would be used to update the routing table and connection pool
+    public static ClusterComposition lookupRoutingTable(
```
Maybe make `LoadBalancer` create an instance of `Rediscovery` with `RoutingSettings`, `Clock`, etc. Otherwise this method has too many parameters. It would also make testing of `LoadBalancer` easier with custom `Rediscovery`.

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/cluster/Rediscovery.java:
```diff
@@ -0,0 +1,95 @@
+                    log.info(
+                            "Server <%s> unable to perform routing capability, dropping from list of routers.",
+                            address );
+                    routingTable.removeRouter( address );
```
Do we still want to remove a router here?

zhenlineo replied:
I am not trying to change existing logic in this PR.

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/cluster/RoutingTable.java:
```diff
@@ -0,0 +1,44 @@
+
+    RoundRobinAddressSet writers();
+
+    BoltServerAddress nextRouter();
```
Could we maybe replace `nextRouter()` and `routerSize()` with `RoundRobinAddressSet routers()`? Then callers would call `routingTable.routers().next()` and `routingTable.routers().size()`.

zhenlineo replied:
It was a on purpose design to make it easier to mock/override to return next router as demands. See the class named `ListBasedRoutingTable`

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/net/ChannelFactory.java:
```diff
@@ -69,7 +65,7 @@ private static void connect( SocketChannel soChannel, BoltServerAddress address,
         catch ( SocketTimeoutException e )
         {
-            throw new IOException( "Timeout " + timeoutMillis + "ms expired", e );
+            throw new ConnectException( "Timeout " + timeoutMillis + "ms expired" + e.getMessage() );
```
I think it is better to keep `e` as an argument, not just `e.getMessage()`

zhenlineo replied:
ConnectException has no constructor with `e` as aurgument

lutovich replied:
Maybe we can keep `IOException`? Javadoc for `ConnectException` says "Typically, the connection was refused remotely (e.g., no process is listening on the remote address/port)" which is a bit different in our case.

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/net/SocketClient.java:
```diff
@@ -120,7 +119,10 @@ public void start()
         {
             logger.debug( "~~ [CONNECT] %s", address );
-            setChannel( ChannelFactory.create( address, securityPlan, timeoutMillis, logger ) );
+            if( channel == null )
```
Could we instead assert that channel is `null` as the first thing in this method and throw `IllegalStateException` if it is not null. This will disallow multiple `#start()` calls.

zhenlineo replied:
The check here is to allow injection of a mocked channel into this class for testing. This null check is for `do not create a new channel if I already give you a fake channel`.

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/security/TLSSocketChannel.java:
```diff
@@ -79,7 +81,21 @@ public TLSSocketChannel( ByteChannel channel, Logger logger, SSLEngine sslEngine
+            {
+                error = e.getCause();
+            }
+            throw new ClientException( "Failed to establish secured connection with the server: " + error.getMessage(),
```
It is probably better to move this `try-catch` block inside `runHandshake()` and just re-throw `ClientException` with whatever `IOException` is thrown as the cause.

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/security/TLSSocketChannel.java:
```diff
@@ -164,6 +180,14 @@ private HandshakeStatus unwrap( ByteBuffer buffer ) throws IOException
         {
+            try
+            {
+                channel.close();
```
Please add a test for this fix

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/security/TLSSocketChannel.java:
```diff
@@ -289,14 +313,19 @@ private HandshakeStatus wrap( ByteBuffer buffer ) throws IOException
-                                    "new network buffer.", curNetSize, netSize, buffer.capacity() ) );
+                    try
+                    {
+                        channel.close();
```
Please add a test for this fix.

lutovich commented on driver/src/test/java/org/neo4j/driver/internal/net/SocketClientTest.java:
```diff
@@ -22,32 +22,69 @@
+    {
+        return Arrays.asList(new Object[][] {
+                { "insecure", SecurityPlan.insecure() },
+                { "encrypted", createEncryptedSecurityPlan() }
```
Can we just use `SecurityPlan.forAllCertificates()` here and declare `#data()` to throw `Exception`?

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/net/SocketClient.java:
```diff
@@ -133,11 +135,7 @@ public void start()
-        catch ( GeneralSecurityException e )
-        {
-            throw new ClientException( "Unable to establish ssl connection with server: " + e.getMessage(), e );
+            throw new ServiceUnavailableException( "Unable to process request: " + e.getMessage(), e );
```
Could a test be added to assert that router is removed when this exception is thrown?

lutovich reviewed (changes requested):
@zhenlineo review completed. Made some comments.

Commit:
Removed the invalid `parameterized` tag in `SocketClient` as the test only tests on a mocked channel
Fixed the bug where the channel might not be closed if socketClient.start() throws exceptions

Added tests to verify channel is closed after construction failure and IO failures for both plaintext and encrypted channel

Commit:
Fix the failing test

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/security/TLSSocketChannel.java:
```diff
@@ -67,10 +68,26 @@ public static TLSSocketChannel create( BoltServerAddress address, SecurityPlan s
+        catch ( SSLHandshakeException e )
+        {
+            throw new ClientException( "Failed to establish secured connection with the server: " + e.getMessage(),
+                    e.getCause() );
```
No reason to get the cause.

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/security/TLSSocketChannel.java:
```diff
@@ -288,16 +345,7 @@ private HandshakeStatus wrap( ByteBuffer buffer ) throws IOException
-                                    "buffer size %s is so big that the application data still cannot fit into the " +
-                                    "new network buffer.", curNetSize, netSize, buffer.capacity() ) );
-                }
+                channelWrite( cipherOut );
```
Let's keep `written == 0` check

Commit:
Fix after review
